docs/developer_guide.md

```diff
@@ -61,6 +61,83 @@ java -jar data-prepper-core-$VERSION.jar pipelines.yaml data-prepper-config.yaml
 
 Optionally add `"-Dlog4j.configurationFile=config/log4j2.properties"` to the command if you would like to pass a custom log4j2 properties file. If no properties file is provided, Data Prepper will default to the log4j2.properties file in the *shared-config* directory.
 
+## Coding Guidance
+
+### Documentation
+
+Documentation is very important for users of Data Prepper and contributors. We are using the
+following conventions for documentation.
+
+1. Document features in markdown. Plugins should have detailed documentation in a `README.md` file in the plugin project directory. Documentation for all of Data Prepper should be in the [docs](../docs) directory.
+2. Provide Javadocs for all public classes, methods, and fields. Plugins need not follow this guidance since their classes are generally not exposed.
+3. Avoid commenting within code, unless it is required to understand that code.
+
+### Code
+
+For the most part, we use common Java conventions. Here are a few things to keep in mind.
+
+1. Use descriptive names for classes, methods, fields, and variables.
+2. Avoid abbreviations unless they are widely accepted
+3. Use final on all variables which are not reassigned
+4. Wildcard imports are not allowed.
+5. Static imports are preferred over qualified imports when using static methods
+6. Prefer creating non-static methods whenever possible. Static methods should generally be avoid as they are often used as a shortcut. Sometimes static methods are the best solution such as when using a builder.
+7. Public utility or “common” classes are not permitted.
+    1. They are fine in test code
+    2. They are fine if package protected
+8. Use Optional for return values if the value may not be present. This should be preferred to returning null.
+9. Do not create checked exceptions, and do not throw checked exceptions whenever possible. In general, if you call a method with a checked exception, you should wrap that exception into an unchecked exception.
```

Comment:
Should this only apply to APIs meant to be overridden? For internal/private helper method, I am not sure if we have added too much restriction on code style here.

Reply:
Yes, that is my intention, and I can clarify it.

Reply:
How is this?

> 9. Do not create checked exceptions, and do not throw checked exceptions from public methods whenever possible. In general, if you call a method with a checked exception, you should wrap that exception into an unchecked exception.
    1. Throwing checked exceptions from private methods is acceptable.

Reply:
Works for me.